File: leetcode/tree/path-sum-ii.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, sum int) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}
	dfs(root, sum, &[]int{}, &result)
	return result
}

func dfs(root *TreeNode, sum int, stash *[]int, result *[][]int)  {
	if root.Left == nil && root.Right == nil {
		if sum == root.Val {
			*stash = append((*stash)[:len(*stash):len(*stash)], root.Val)
			*result = append(*result, *stash)
			return
		}
	}
	if root.Left != nil {
		*stash = append((*stash)[:len(*stash):len(*stash)], root.Val)
		dfs(root.Left, sum-root.Val, stash, result)
		*stash = (*stash)[:len(*stash)-1]
	}
	if root.Right != nil {
		*stash = append((*stash)[:len(*stash):len(*stash)], root.Val)
		dfs(root.Right, sum-root.Val, stash, result)
		*stash = (*stash)[:len(*stash)-1]
	}
}

func buildTree(nums []int, i int) *TreeNode {
	if nums == nil {
		return nil
	}
	if i > len(nums) {
		return nil
	}
	if nums[i-1] == 0 {
		return nil
	}
	root := &TreeNode{Val:nums[i-1]}
	root.Left = buildTree(nums, 2*i)
	root.Right = buildTree(nums, 2*i+1)
	return root
}

func main() {
	nums := []int{5,4,8,11,0,13,4,7,2,0,0,0,0,5,1}
	root := buildTree(nums, 1)
	//head := &ListNode{}
	fmt.Println(pathSum(root, 22))
}

File: python/tree/path-sum-ii.py
import copy

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.stash = []

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        result = []
        self.help(root, sum, result)
        return result

    def help(self, root, sum, result):
        self.stash.append(root.val)
        if root.left is None and root.right is None:
            if sum == root.val:
                result.append(self.stash)
            return
        if root.left is not None:
            self.help(root.left, sum - root.val, result)
        if root.right is not None:
            self.help(root.right, sum - root.val, result)
        self.stash.pop()

    def toBST(self, arr, i=1):
        if not arr:
            return None
        if i > len(arr):
            return None
        if arr[i-1] is None:
            return None
        root = TreeNode(arr[i-1])
        root.left = self.toBST(arr, 2*i)
        root.right = self.toBST(arr, 2*i+1)
        return root

if __name__ == '__main__':
    origin = [5, 4, 8, 11, None, 13, 4, 7, 2, None, None, 5, 1]
    s = Solution()
    a = s.toBST(origin)
    print(s.pathSum(a, 22))
